### PR TITLE
fuse tril scale softmax dropout

### PR DESCRIPTION
**Pull request by guo-ran**

### CODE REVIEW

File: oneflow/user/ops/fused_scale_tril_softmax_mask_and_scale_op.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+namespace {
+
+REGISTER_USER_OP("fused_tril_scale_softmax_mask_and_scale")
+    .Input("x")
+    .Input("mask")
+    .Output("y")
+    .Output("softmax_y")
+    .Attr<int64_t>("diagonal")
+    .Attr<double>("floating_fill_value", 0)
```
Comment (liujuncheng) [inline oneflow/user/ops/fused_scale_tril_softmax_mask_and_scale_op.cpp:28]: floating_tril_fill_value?

File: oneflow/user/ops/fused_scale_tril_softmax_mask_and_scale_op.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    .Attr<int64_t>("integer_fill_value", 0)
+    .Attr<bool>("is_floating_fill_value", false)
+    .Attr<double>("floating_prologue_scale_value", 1)
+    .Attr<int64_t>("integer_prologue_scale_value", 1)
+    .Attr<bool>("is_floating_prologue_scale_value", false)
+    .Attr<float>("epilogue_scale")
```
Comment (liujuncheng) [inline oneflow/user/ops/fused_scale_tril_softmax_mask_and_scale_op.cpp:34]: epilogue_scale_value 或者 上面的 `integer_prologue_scale_value` 不加 value

File: oneflow/user/ops/fused_scale_tril_softmax_mask_and_scale_op.cpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    .SetTensorDescInferFn([](user_op::InferContext* ctx) -> Maybe<void> {
+      *ctx->TensorDesc4ArgNameAndIndex("y", 0) = *ctx->TensorDesc4ArgNameAndIndex("x", 0);
+      *ctx->TensorDesc4ArgNameAndIndex("softmax_y", 0) = *ctx->TensorDesc4ArgNameAndIndex("x", 0);
```
Comment (liujuncheng) [inline oneflow/user/ops/fused_scale_tril_softmax_mask_and_scale_op.cpp:37]: 提取`const TensorDesc* x = ctx->TensorDesc4ArgNameAndIndex("x", 0)`

File: oneflow/user/ops/fused_scale_tril_softmax_mask_and_scale_op.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      return Maybe<void>::Ok();
+    })
+    .SetInputArgModifyFn([](user_op::GetInputArgModifier GetInputArgModifierFn,
```
Comment (liujuncheng) [inline oneflow/user/ops/fused_scale_tril_softmax_mask_and_scale_op.cpp:40]: softmax_y也不需要梯度吧

File: oneflow/user/ops/fused_scale_tril_softmax_mask_and_scale_op.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                            const user_op::UserOpConfWrapper&) {
+      user_op::InputArgModifier* mask_modifier = GetInputArgModifierFn("mask", 0);
+      CHECK(mask_modifier != nullptr);
+      mask_modifier->set_requires_grad(false);
+    })
+    .SetGetSbpFn([](user_op::SbpContext* ctx) -> Maybe<void> {
+      const user_op::TensorDesc& x_tensor = ctx->LogicalTensorDesc4InputArgNameAndIndex("x", 0);
+      FOR_RANGE(int64_t, axis, 0, x_tensor.shape().NumAxes()) {
```
Comment (liujuncheng) [inline oneflow/user/ops/fused_scale_tril_softmax_mask_and_scale_op.cpp:48]:
```suggestion
      FOR_RANGE(int64_t, axis, 0, x_tensor.shape().NumAxes() - 2) {
```

File: oneflow/user/ops/fused_scale_tril_softmax_mask_and_scale_op.cpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    .Attr<int64_t>("diagonal")
+    .Attr<double>("floating_epilogue_scale_value", 1)
+    .Attr<int64_t>("integer_epilogue_scale_value", 1)
+    .Attr<bool>("is_floating_epilogue_scale_value", false)
+    .Attr<float>("prologue_scale")
+    .SetTensorDescInferFn([](user_op::InferContext* ctx) -> Maybe<void> {
+      const user_op::TensorDesc* softmax_y_desc = ctx->TensorDesc4ArgNameAndIndex("softmax_y", 0);
+      const user_op::TensorDesc* dy_desc = ctx->TensorDesc4ArgNameAndIndex("dy", 0);
+      user_op::TensorDesc* dx_desc = ctx->TensorDesc4ArgNameAndIndex("dx", 0);
+      CHECK(*dy_desc == *softmax_y_desc);
```
Comment (liujuncheng) [inline oneflow/user/ops/fused_scale_tril_softmax_mask_and_scale_op.cpp:73]: 小心dynamic

File: oneflow/user/ops/fused_scale_tril_softmax_mask_and_scale_op.cpp [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+      *dx_desc = *dy_desc;
+      return Maybe<void>::Ok();
+    })
+    .SetGetSbpFn([](user_op::SbpContext* ctx) -> Maybe<void> {
+      const user_op::TensorDesc& y_tensor = ctx->LogicalTensorDesc4InputArgNameAndIndex("y", 0);
+      FOR_RANGE(int64_t, axis, 0, y_tensor.shape().NumAxes()) {
```
Comment (liujuncheng) [inline oneflow/user/ops/fused_scale_tril_softmax_mask_and_scale_op.cpp:79]:
```suggestion
      FOR_RANGE(int64_t, axis, 0, y_tensor.shape().NumAxes() - 2) {
```

File: oneflow/user/kernels/fused_tril_scale_softmax_mask_and_scale_kernel.cu [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+
+template<typename T>
+T GetAttrVal(bool is_floating_val, double floating_value, int64_t integer_value) {
+  return is_floating_val ? static_cast<T>(floating_value) : static_cast<T>(integer_value);
+}
+
+template<>
+half GetAttrVal<half>(bool is_floating_val, double floating_value, int64_t integer_value) {
+  return is_floating_val ? __float2half(floating_value) : __float2half(integer_value);
+}
```
Comment (liujuncheng) [inline oneflow/user/kernels/fused_tril_scale_softmax_mask_and_scale_kernel.cu:29]: 应该不需要

File: oneflow/user/kernels/fused_tril_scale_softmax_mask_and_scale_kernel.cu [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+template<typename SRC>
+struct TrilScaleFetch {
+  template<typename DST, int N>
+  __device__ void fetch(DST* dst, int32_t row, int64_t col) {
```
Comment (liujuncheng) [inline oneflow/user/kernels/fused_tril_scale_softmax_mask_and_scale_kernel.cu:34]:
```suggestion
  __device__ void fetch(DST* dst, int64_t row, int64_t col) {
```

File: oneflow/user/kernels/fused_tril_scale_softmax_mask_and_scale_kernel.cu [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    int32_t tril_row = row % tril_num_rows;
```
Comment (liujuncheng) [inline oneflow/user/kernels/fused_tril_scale_softmax_mask_and_scale_kernel.cu:35]:
```suggestion
    int64_t tril_row = row % tril_num_rows;
```

File: oneflow/user/kernels/fused_tril_scale_softmax_mask_and_scale_kernel.cu [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+  SRC fill;
+  SRC scale;
+};
+
+template<typename DST>
+struct MaskAndScaleStore {
+  template<typename SRC, int N>
+  __device__ void store(const SRC* src, int64_t row, int64_t col) {
+    cuda::softmax::Pack<DST, N> softmax_out_pack;
+    cuda::softmax::Pack<DST, N> pack;
```
Comment (liujuncheng) [inline oneflow/user/kernels/fused_tril_scale_softmax_mask_and_scale_kernel.cu:65]: dst_pack

File: oneflow/user/kernels/fused_tril_scale_softmax_mask_and_scale_kernel.cu [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+    int32_t tril_row = row % tril_num_rows;
+    bool need_fetch = (col <= (tril_row + diagonal));
+    cuda::softmax::Pack<SRC, N> pack;
+    if (need_fetch) {
+      int64_t offset = row * row_size + col;
+      pack.storage = *reinterpret_cast<const cuda::softmax::PackType<SRC, N>*>(src + offset);
+    }
+#pragma unroll
+    for (int i = 0; i < N; ++i) {
+      if (col + i > tril_row + diagonal) {
```
Comment (liujuncheng) [inline oneflow/user/kernels/fused_tril_scale_softmax_mask_and_scale_kernel.cu:44]: `tril_row + diagonal` 是重复的表达式

**Review by liujuncheng — APPROVED** [review]
